### PR TITLE
policy: Add Port Range Support for Policies Part 2/3

### PR DESCRIPTION
**Pull request by nathanjsweet**
This PR prepares the policy engine for adding port ranges
by enabling the underlying userspace cache to calculate
insertion, deletion, and lookups with port ranges, as well
as adding unit tests to ensure that the logic works. It does
not add support for adding policy port ranges at the API
level that will be addressed in the final PR.

The Policy CRD is modified by this PR without
supporting port ranges at the policy repository level
(this will be added in the final PR). This has to be done 
because the "PortProtocol" struct is shared by both 
the CRD (aka the API level) and the L4Filter struct
(aka the cache level).

See commits for details.

### CODE REVIEW

File: pkg/policy/mapstate.go [hunk 1]
```diff
@@ -284,6 +284,29 @@ func (e *MapStateEntry) HasDependent(key Key) bool {
...
+// HasSameOwners returns true if both MapStateEntries
+// have the same owners as one another (which means that
+// one of the entries is redundant).
+func (e *MapStateEntry) HasSameOwners(bEntry *MapStateEntry) bool {
+	if e == nil && bEntry == nil {
+		return true
+	}
+	if e == nil || bEntry == nil {
+		return false
+	}
```
Comment (tommyp1ckles) [inline pkg/policy/mapstate.go:293]: you could return false early here if `len(e.owner) != len(bEntry.owner)`

**Review by tommyp1ckles — APPROVED** [review]
endpoint changes lgtm

**Review by kaworu — APPROVED** [review]
Hubble changes LGTM

**Review by learnitall — APPROVED** [review]
API changes LGTM!

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml [hunk 2]
```diff
@@ -548,15 +548,15 @@ spec:
                               description: PortProtocol specifies an L4 port with
                                 an optional transport protocol
                               properties:
+                                endPort:
+                                  description: EndPort can be an L4 port number.
```
Comment (doniacld) [inline pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml]: Can it be something else than a L4 port number?

**Review by doniacld — APPROVED** [review]
One nit question, otherwise LGTM! 🫰

File: pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml (hunk 2 above)
Comment (nathanjsweet) [reply pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml]: No. Good point.

**Pull request merged**